pip/pip-299.md

```diff
@@ -0,0 +1,102 @@
+
+# Motivation
+
+Since the frequent persistence of `Individual Deleted Messages` will magnify the amount of BK Written and increase the latency of ack-response, the Broker does not immediately persist it when receiving a consumer's acknowledgment but persists it regularly. 
+
+The data of cursor metadata is recommended to be less than 5MB; if a subscription's `Individual Deleted Messages` data is too large to persist, as the program grows for a long time, there will be more and more non-persistent data. Eventually, there will be an unacceptable amount of repeated consumption of messages when the Broker restarts.
+
+# Goal
+
+## In Scope
+
+To avoid repeated consumption due to the cursor metadata being too large to persist.
+
+## Out of Scope
+
+This proposal will not care about this scenario: if so many producers make the metadata of cursor `pulsar.dedup` cannot persist, the task `Take Deduplication Snapshot` will be in vain due to the inability to persist.
+
+# High-Level Design
+
+**1.Max ack state to persist**
+Provide a new config named `managedLedgerMaxAckStateInBytesToPersist`, persisting only part of the data according to the following priorities.
+- Subscription Properties
+- Last sequence ID of producers
+- Individual Deleted Message
+
+For example, we set `managedLedgerMaxAckStateInBytesToPersist` to `5m`, now Subscription Properties costs `1m`, the Last sequence ID of producers costs `0bytes`, and Individual Deleted Message costs `10m`. The Broker will persist all Subscription Properties(`1m`) and a part of Individual Deleted Message(`4m`). The extra part of the Individual Deleted Message(`6m`) will not persist.
+
+Note: `managedLedgerMaxAckStateInBytesToPersist` and `managedLedgerMaxUnackedRangesToPersist` are conflicting configurations. If the two configurations are set at the same time, only `managedLedgerMaxAckStateInBytesToPersist` takes effect.
+
+**Pause delivery messages if reaches the limitation**
+Cache the size of the cursor metadata in memory when doing persistent data to BK. We call the cache name `cursorMetadataSizeInMemoryBytes.` 
+
+Note:
+- Since we will not update `cursorMetadataSizeInMemoryBytes` each time acknowledgment, `cursorMetadataSizeInMemoryBytes` is not a real-time value.
+- The delayed messages will also not be redelivered after reaching the limitation.
+
+Provide a new config named `dispatcherPauseOnAckPersistentStateEnabled,` stuck delivery messages to clients if the size of the cursor metadata could not persist due to it being too large.
+
+
+
+# Detailed Design
+### Public API
+
+**broker.conf**
+```java
+/** 
+ * Max bytes size of the cursor metadata. If it reaches this limitation, Pulsar will not try to persist cursor metadata(including ack state).
+ * The value of this property should be between 1K and 5M. The default value is 5M.
+ */
+long managedLedgerMaxAckStateInBytesToPersist;
```

Comment:
Is it better to move this change to a separate proposal?

- This proposal: Stop dispatch messages if the individual acks will be lost in the persistent storage
- Another proposal: Support size-based ack state persistence, because the ranges-bases configuration is not easy to be understand by users.

Reply:
Good suggestion. Current PIP only contains the design of stop dispatch messages if the individual acks will be lost in the persistent storage now. And I will submit another PIP to do the second part after the current PIP is done.